### PR TITLE
os/ObjectStore: implement more efficient get_encoded_bytes() 

### PR DESCRIPTION
**Pull request by branch-predictor**
Previous version explicitly encoded entire transaction just to get
its length, which resulted in encoding it twice. New method adds
encoded_size() methods to ghobject_t and coll_t, which return their
encoded size without explicit encoding, and in turn, reducing memory
manager pressure and CPU time spent. Added a test so a change
to ghobject_t and/or coll_t encoding should trigger test failure.

Signed-off-by: Piotr Dałek <piotr.dalek@ts.fujitsu.com>

### CODE REVIEW

File: src/os/ObjectStore.h [hunk 1]
```diff
@@ -808,7 +808,37 @@ class ObjectStore {
...
+        // coll_index size, object_index size and sizeof(transaction_data)
+        // all here, so they may be computed at compile-time
+        size_t final_size = sizeof(__u32) * 2 + sizeof(data);
+
+        // coll_index second and object_index second
+        final_size += (coll_index.size() + object_index.size()) * sizeof(__le32);
+
+        // coll_index first
+        for (auto p = coll_index.begin(); p != coll_index.end(); ++p) {
+          final_size += p->first.encoded_size();
```
Comment (yuyuyu101) [inline src/os/ObjectStore.h]: encoded_size could be cached or calculated when global init?

Comment (branch-predictor) [inline src/os/ObjectStore.h]: I don't think so.

Comment (yuyuyu101) [inline src/os/ObjectStore.h]: oh, sorry. I'm wrong

**Comment by yuyuyu101** [comment]
A short unittest benchmark result?

**Comment by branch-predictor** [comment]
From what I've measured, for typical write ops the median time for new get_encoded_bytes is 14ns, and for old - 1613ns (which is not surprising considering what old code actually did).

**Comment by branch-predictor** [comment]
Added new unittest benchmark:
```
[ RUN      ] Transaction.GetNumBytesBenchLegacy
get_encoded_bytes_test: 2500000 encodes in 3.988424
[       OK ] Transaction.GetNumBytesBenchLegacy (3988 ms)
[ RUN      ] Transaction.GetNumBytesBenchCurrent
get_encoded_bytes: 2500000 encodes in 0.040090
[       OK ] Transaction.GetNumBytesBenchCurrent (40 ms)
```

**Pull request merged**